### PR TITLE
do not let Body read beyond its length

### PR DESCRIPTION
**Pull request by jbr**
this logic is currently handled in several places in async-h1, but it should be the responsibility of Body to ensure that if it has a set length, it will not read beyond that length.

I believe this is semver-patch, unless we consider "reading past set body length" intended behavior

### CODE REVIEW

File: src/body.rs [hunk 1]
```diff
@@ -459,15 +468,25 @@ impl AsyncRead for Body {
         cx: &mut Context<'_>,
         buf: &mut [u8],
     ) -> Poll<io::Result<usize>> {
-        Pin::new(&mut self.reader).poll_read(cx, buf)
+        let mut buf = match self.length {
+            None => buf,
+            Some(length) if length == self.bytes_read => return Poll::Ready(Ok(0)),
```
Comment (jbr) [inline src/body.rs:473]: this line is an optional optimization and could be removed if it makes the code more readable to do so. i was on the fence about this tradeoff

File: src/body.rs [hunk 2]
```diff
@@ -459,15 +468,25 @@ impl AsyncRead for Body {
...
+            Some(length) => {
+                let max_len = (length - self.bytes_read).min(buf.len());
+                &mut buf[0..max_len]
```
Comment (Fishrock123) [inline src/body.rs:476]: Will this panic if the body has more bytes than it specifies?

Comment (jbr) [reply src/body.rs:476]: nope, there's a test for that. we just won't read them from the underlying Read, which is the correct behavior for keepalive

Comment (jbr) [reply src/body.rs:476]: this line limits the length of the mutable slice that we pass into the inner read implementation, effectively telling it "give me a maximum of `max_len` bytes

**Pull request merged**